### PR TITLE
[ValueTracking] Infer relationship for the select with ICmp

### LINKED ISSUE #54735: [optimization] gcc generate better code than clang base on value range propagation
after fix the case https://github.com/llvm/llvm-project/issues/54132, the clang still need be improved for some more complex case （see detail on **https://godbolt.org/z/xb9Mo85WM** ）, such as:

```
#include <stdlib.h>

int my_abs (int x, int y) {
  return x>y ? abs (x-y+1): 0;
}
```
* clang's code：
```
my_abs: // @my_abs
  subs w8, w0, w1
  b.le .LBB0_2
  mvn w9, w8
  cmn w8, #1
  csinc w0, w9, w8, lt
  ret
.LBB0_2:
  mov w0, wzr
  ret
```
* gcc's code:
```
my_abs:
  subs w0, w0, w1
  csinc w0, wzr, w0, le
  ret
```

### PR DESCRIPTION
**Pull request by vfdff**
x -nsw y < C is false when x > y and C <= 0
Alive2 proof for sgt, sge, ugt, uge : https://alive2.llvm.org/ce/z/tupvfi

Fixes https://github.com/llvm/llvm-project/issues/54735

### CODE REVIEW

File: llvm/test/Transforms/InstSimplify/select-icmp.ll
Comment (goldsteinn) [inline llvm/test/Transforms/InstSimplify/select-icmp.ll:33]: Please split to two commits, 1st is tests before the change, 2nd is change (with updated tests). Allows us to see the actual affects of the patch.

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (goldsteinn) [inline llvm/lib/Analysis/InstructionSimplify.cpp:4561]: You can fallback on knownbits. See `InstCombeCalls::getKnownSign`.
In fact, I'd say a better approach would be to port `getKnownSign` to a mutually accessible location (maybe `ValueTracking` then just that here.

**Review by nikic — CHANGES_REQUESTED** [review]
I believe this needs to be handled in isImpliedCondition. There is already existing code that will simplify implications from dominating conditions.

Though really, this seems like something that ConstraintElimination should handle, not InstSimplify.

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:8291]:
```suggestion
    // x -nsw y is non-negative when x >= y or negative when x < y
```
You match only the sub here, not an extra +1, so the comment should match that.

**Review by nikic — COMMENTED** [review]
Please add an alive2 proof to the patch description. This should also needs additional tests (preferably more minimal than the example you added), e.g. that this does not fold without nsw, wrong constant, etc.

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8291]: hi, @nikic,
   the **extra +1** is matched with the **match(R1, m_AllOnes())**.

**Comment by vfdff** [comment]
> Please add an alive2 proof to the patch description. This should also needs additional tests (preferably more minimal than the example you added), e.g. that this does not fold without nsw, wrong constant, etc.

Thanks, Added [alive2 proof link](https://alive2.llvm.org/ce/z/aGAC3j) and 2 extra negative test cases.

**Comment by vfdff** [comment]
ping ?

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:8294]: It is incorrect that `X <s Y` implies `X -nsw Y <s -1`.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8294]: Oh, you are right. 
I missed the big premise, and valuetracking can be used except `abs`, so `X -nsw Y <s -1` and `X -nsw Y <=s -1` is unequivalent.

Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp]: Thanks, now I restrict it `x -nsw y <= -1 <==> x < y`.
And add extra match pattern `x > y and C <= 0 ==> x -nsw y < C is false` to fix #54735

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8288]: you'll see now we also can't optimize the following case
```
#include <stdlib.h>
int test_slt (int x, int y) {
  return x<y ? abs (x-y+1): 0;   // here `x-y+1 <= 0`
}
```

Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8271]: hi @nikic, should I remove this branch ?

Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8271]: ping ?

**Comment by vfdff** [comment]
revert the change for `x -nsw y <= -1 <==> x < y` because `sle -1 will get canonicalized to slt 0`, so should consider other way to address that.

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:8294]: Can we handle more predicates?

File: llvm/test/Transforms/InstSimplify/select-icmp.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstSimplify/select-icmp.ll:5]: Please drop `noundef` attributes.

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8294]: hi @dtcxzyw
Because different predicates have some different logics to address,
I think it is fine to keep this PR to fix the issue in the comment link,
and I'll try to expand some others implemented separately.

**Comment by vfdff** [comment]
ping ?

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:8304]: (nit) `if(!isImplied...)`

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:8297]: They don't seem all that different (for `sge`, `uge`, and `ugt` that is).
https://alive2.llvm.org/ce/z/aNpE5W

You only need the `C <= 0` for signed predicates.
For signed you need `nsw`. For unsigned you need `nuw`.

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:8297]: Should also handle the less-than variant no?

Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8297]: hi @goldsteinn
I reverted the change for less-than because the commit https://github.com/llvm/llvm-project/pull/66668#pullrequestreview-1682554666, so I think it may be addressed with some other method.
If you don't object, I think it is fine to keep this PR to fix the issue in the comment link.

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8297]: Thanks, I'll extend this

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8304]: the return value of `isImpliedCondMatchingOperands` may be **false**, **true** or **null**, so `!isImplied...` is not equal to `isImplied... == false`

**Comment by vfdff** [comment]
ping ?

**Comment by vfdff** [comment]
rebase to fix the conflict to commit 460faa0c87

**Comment by vfdff** [comment]
ping ?

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:8377]:
```suggestion
    if (match(R1, m_NonPositive()) &&
```

`CI` is unused.

**Review by dtcxzyw — APPROVED** [review]
The implementation LGTM. But it would be good to support more predicates (slt/ult).

Please wait for additional approval from other reviewers.

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8377]: Apply your comment, thanks

**Comment by vfdff** [comment]
> The implementation LGTM. But it would be good to support more predicates (slt/ult).
> 
> Please wait for additional approval from other reviewers.
Thanks for your notification, I'll follow to address `slt/ult` after this PR is accepted.

**Comment by vfdff** [comment]
hi @nikic, would you help to give some suggestion on this PR ?

**Comment by vfdff** [comment]
ping ?

**Comment by vfdff** [comment]
If there's no objection, I'll take a try for this commit in the next week?

**Comment by dtcxzyw** [comment]
Ping @nikic @goldsteinn

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:8341]: Just based on the comment, I would have expected `R1` to be non-negative?

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:8344]: you could handle `SLT`/`SLE`/`ULT`/`ULE` as `true` here no?

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8893]: Thanks, I update the comment, also rebase to top as the upstream has delete the API **areMatchingOperands** since commit 01d8e1ca

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8896]: Ok, thanks for your idea, I'll address that with a separate PR later, and this PR is just intend to fix  [#54735](https://github.com/llvm/llvm-project/issues/54735)

Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8895]: hi @dtcxzyw, @goldsteinn 
 I add the addressing for **SLT/SLE/ULT/ULE** now, thanks

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:8907]: Both here and above, instead of m_Value(X)/m_Value(Y), how about m_Specific(L0)/m_Specific(L1)?

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:8892]: Here and below, don't think you need `nuw` for the sub. Its implied by LPred.

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:8892]: https://alive2.llvm.org/ce/z/yQTr5f

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:8892]: Actually I misspoke, but I don't think your proofs are correct. The `nuw` on the `sub nuw i8 0, %add` forces `%add` to be 0. Likewise the `sub nuw i8 %sub, %c` essentially forces the condition you want, not the dominating predicates.

Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8892]: Thanks @goldsteinn , If this only valid for 0 with nuw, so the transformation of unsigned should be deleted (only keep it for signed types) ? or do you have other suggestion ？

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:8892]: Yeah, unsigned doesn't make sense here. The is really saying "X - Y must be positive if X >= Y and no overflow". That only really makes sense in the context of signed comparison.

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8904]: Apply your comment, thanks

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp]: Thanks for the confirmation, and delete the parts of unsigned.

File: llvm/test/Transforms/InstSimplify/select-icmp.ll
Comment (goldsteinn) [inline llvm/test/Transforms/InstSimplify/select-icmp.ll:373]: You can drop the unsigned tests.

Comment (vfdff) [reply llvm/test/Transforms/InstSimplify/select-icmp.ll:245]: Apply your comments, thanks @goldsteinn

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:8895]: Is this right for `sge` if `C == 0`? Your proofs checkout but afaict its b.c your proof is using `X != -X` to handle the false case which has problems with this case...

Comment (vfdff) [reply llvm/lib/Analysis/ValueTracking.cpp:8895]: Does  the link https://alive2.llvm.org/ce/z/N4uvTh proof it is valid for **sge** with **C == 0** ?

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:8895]: I made a mistake think this is okay.

**Comment by goldsteinn** [comment]
LGTM. Please wait for an additional approval

**Comment by vfdff** [comment]
> LGTM. Please wait for an additional approval
Thank very much for your reviewing

**Comment by vfdff** [comment]
ping ?

**Comment by vfdff** [comment]
This PR has no update for a long time, so try  rebasing to top upstream to see if there's any conflict.

**Comment by vfdff** [comment]
ping
can I try to land this PR as there is no new comment for a long time ?

**Comment by goldsteinn** [comment]
LGTM, should be fine to push.

**Review by goldsteinn — APPROVED** [review]

**Comment by vfdff** [comment]
Thanks very much

**Pull request merged**
